### DISCUSSION TITLE
How do I inherit a class?

### DISCUSSION THREAD

**Discussion by ravener** [gircore/gir.core#1121] [Q&A] [answered]
How does inheritance work? Specifically calling the parent's constructor.

I want to subclass `Adw.ApplicationWindow` for example, since this wants the application instance in its constructor how would I pass it, I was not able to use the `base()` properly, and I can't use `.New()` when subclassing it.

Goal is something like this

```cs

class MyWindow : Adw.ApplicationWindow
{
    public MyWindow(Gtk.Application app) : base(app)
    {
        // Do stuff
    }
}
```

**Comment by ravener** [comment]
Well some progress, I did this

```cs

class MyWindow : Adw.ApplicationWindow
{
    public MyWindow(Gtk.Application app) : base(app.Handle, false)
    {
        // Do stuff
    }
}
```

That fixed the constructor idea, but now when I `new MyWindow(app)` and `window.Present()` it crashes with

```
Gtk-CRITICAL **: 20:38:56.207: gtk_window_present_with_time: assertion 'GTK_IS_WINDOW (window)' failed
```

**Comment by ravener** [comment]
okay after looking closer at how Adw.Application.New works I realized this is the correct way:

```cs

class MyWindow : Adw.ApplicationWindow
{
    public MyWindow(Gtk.Application app) : base(New(app).Handle, false)
    {
        // Do stuff
    }
}
```

This works, though I'm unsure if that's the best practice, how are others dealing with it?

**Comment by ravener** [comment, marked as answer]
Seems like the alternative is to not initialize the constructor but then later when presenting the window use `application.AddWindow`

According to Gtk docs this is equivalent to setting the application property of the window so that works and looks better.

**Comment by badcel** [comment]
To create a proper subclass which is known to the GObject typesystem you would need to call into [this constructor](https://github.com/gircore/gir.core/blob/466825a547d8498e071ea0563c47abd98c9f42e9/src/Libs/GObject-2.0/Public/Object.cs#L37)

In your case this would be [this](https://gircore.github.io/api/Adw.ApplicationWindow.html#Adw_ApplicationWindow__ctor_GObject_ConstructArgument___) constructor.

In general I would advice against subclasses as GTK itself is trying to avoid it. In general for a C# programmer I would only recommend subclasses if there are specific arguments for it (similar to GTK) as inheritance chains are not very flexible and tend to lead to complex / harder to maintain code.

**Reply by ravener** [reply]
Yeah, I don't like dealing with GObject types, it's just a common pattern to separate the main window to a class extending Adw.ApplicationWindow, I've seen it done everywhere from Vala to Rust, so at least for that class.
